moveit_core/planning_scene/test/test_multi_threaded.cpp

```diff
@@ -117,23 +117,24 @@ TEST_P(CollisionDetectorTests, Threaded)
 
   for (unsigned int i = 0; i < THREADS; ++i)
   {
-    moveit::core::RobotState* state = new moveit::core::RobotState(planning_scene_->getRobotModel());
+    moveit::core::RobotState state{ planning_scene_->getRobotModel() };
     collision_detection::CollisionRequest req;
-    state->setToRandomPositions();
-    state->update();
-    states.push_back(moveit::core::RobotStatePtr(state));
-    collisions.push_back(runCollisionDetection(0, 1, planning_scene_.get(), state));
+    state.setToRandomPositions();
+    state.update();
+    collisions.push_back(runCollisionDetection(0, 1, *planning_scene_, state));
+    states.push_back(std::move(state));
   }
 
   for (unsigned int i = 0; i < THREADS; ++i)
-    threads.push_back(new std::thread(
-        std::bind(&runCollisionDetectionAssert, i, TRIALS, planning_scene_.get(), states[i].get(), collisions[i])));
+    threads.emplace_back(std::thread([i, &scene = *planning_scene_, &state = states[i], expected = collisions[i]] {
```

Comment:
I've noticed that there are several occasions where pointers are dereferenced like here. Is there a particular reason instead of just dereferencing them in the function call? Just curious @v4hn

Reply:
I thinks, it's just a matter of what is actually captured, e.g. the full `states` array or a single state (reference).

Reply:
Yeah, with states it makes sense to me, with pointers like `planning_scene_` I'm not sure. At least I haven't seen this with local shared pointers afaik, this was my only concern at first.

Reply:
> Yeah, with states it makes sense to me, with pointers like planning_scene_ I'm not sure. At least I haven't seen this with local shared pointers afaik which was my first concern.

I did that for multiple reasons:
- Restrict scope as Robert explained above
- clang-tidy does it automatically :smiley: 
- Never share ownership with methods that don't need it. It's a rather poisonous pattern in the whole code-base to pass shared pointers for everything, even if no ownership is transferred. It's one of the various things @tylerjw complains about a lot :cat2: 
- Dereferencing pointers when they are captured instead of used makes sure you actually get a traceable error at point of definition and not just at point of use. That pattern can help avoid multiple hours of debugging when `std::function`s are copied around.